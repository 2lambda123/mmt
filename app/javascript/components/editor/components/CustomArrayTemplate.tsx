/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react/static-property-placement */
/* eslint-disable react/jsx-no-useless-fragment */
import { observer } from 'mobx-react'
import React, { MouseEventHandler, ReactNode } from 'react'
import { Row, Col } from 'react-bootstrap'
import { retrieveSchema } from 'react-jsonschema-form/lib/utils'
import MetadataEditor from '../MetadataEditor'

type Element = {
  index: number,
  key: string,
  children: ReactNode,
  className: string,
  onDropIndexClick: (index:number) => MouseEventHandler<HTMLButtonElement>
}

type CustomArrayTemplateProps = {
  items: Element[]
  canAdd: boolean
  onAddClick: (event: object) => void
  title: string
  schema: {
    description: string
    items: {
      default?: string
      definitions: any
    },
    definitions: unknown
  }
  uiSchema: unknown,
  options: {
    editor: MetadataEditor
  }
}

class CustomArrayTemplate extends React.Component<CustomArrayTemplateProps, never> {
  static defaultProps: { options: { editor: MetadataEditor } }
  private scrollRef: Array<React.RefObject<HTMLDivElement>>
  constructor(props: CustomArrayTemplateProps) {
    super(props)
    const { items, schema, options } = this.props
    const { editor } = options
    this.scrollRef = []
    const { definitions } = editor.fullSchema
    const itemsSchema = retrieveSchema(schema.items, definitions)
    const { type } = itemsSchema
    if (type && type === 'string') {
      schema.items.default = ''
    }
    items.forEach(() => {
      this.scrollRef.push(React.createRef())
    })
  }
  componentDidUpdate(): void {
    const { options } = this.props
    const { editor } = options
    // Keep this setTimeout at 200 ms. If it is anything less than 200ms it will set the
    // FocusArrayField to null before the page renders causing a console error.
    if (editor.focusArrayField) {
      setTimeout(() => {
        editor.setArrayField(-1)
      }, 300)
    }
  }

  render() {
    const {
      items, canAdd, title, onAddClick, schema, uiSchema, options
    } = this.props
    const { editor } = options
    const uiTitle = uiSchema['ui:title']
    const uiClassNames = uiSchema['ui:classNames']
    items.forEach(() => {
      this.scrollRef.push(React.createRef())
    })
    if (editor?.focusArrayField >= 0) {
      setTimeout(() => {
        this.scrollRef[editor.focusArrayField]?.current?.scrollIntoView({ behavior: 'smooth' })
      }, 200)
    }

    let titleName = title
    if (uiTitle) {
      titleName = uiTitle
    } else if (title.charAt(title.length - 1) === 's') {
      titleName = title.slice(0, -1)
    }
    return (
      <div data-testid="custom-array-template">
        {uiClassNames ? (
          <div className={uiClassNames}>{titleName}</div>
        ) : (
          <div className="title" style={{ borderBottom: 'solid 2px rgb(240,240,240', paddingBottom: 14, paddingTop: 14 }}>
            {titleName}
          </div>
        )}
        <p>
          {schema.description}
        </p>
        <br />

        {items && items.map((element: Element, index: number) => (
          <div
            key={element.key}
            className={element.className}
            style={{
              borderLeft: 'solid 6px rgb(240,240,240', borderBottom: 'solid 2px rgb(240,240,240)', marginBottom: 25, paddingLeft: 10
            }}
            ref={this.scrollRef[index]}
          >
            <div className="custom-array-template-remove-btn">
              <div style={{
                display: 'flex', alignItems: 'center', justifyContent: 'space-between', marginRight: -12, borderBottom: 'solid 2px rgb(240,240,240)', marginBottom: 10, paddingBottom: 10
              }}
              >
                <h6 style={{ height: 14 }}>
                  <b>{titleName}</b>
                  <span style={{ color: 'gray', marginLeft: 5, paddingTop: 20 }}>
                    {items.length > 0 && `(${index + 1} of ${items.length})`}
                  </span>
                </h6>
                <button className="btn" onClick={element.onDropIndexClick(element.index)} type="button">
                  <span style={{ color: 'red' }}>
                    <i className="fa-solid fa-circle-minus fa-2xl" />
                  </span>
                  {' '}
                  <span style={{ color: 'red', fontWeight: 'bold' }}>
                    {' '}
                    Remove
                  </span>
                </button>

              </div>
              <Row className="mb-2  d-flex align-items-center">

                <Col xs="12" lg="12" key={element.key} className={element.className} data-testid="custom-array-element">
                  {element.children}
                </Col>
              </Row>
            </div>
          </div>
        ))}

        {
          (items.length === 0 && canAdd) && (
            <div className="custom-array-template-add-btn" data-testid="custom-array-template-add-btn" style={{ marginTop: -30 }}>
              <button
                className="btn"
                type="button"
                onClick={(e) => {
                  onAddClick(e)
                  editor.setArrayField(items.length)
                }}
              >
                <span style={{ color: 'rgb(31, 107, 162)', marginLeft: '-12px' }}>
                  <i className="fa-solid fa-circle-plus fa-2xl" />
                </span>
                {' '}
                <span style={{ color: 'rgb(31, 107, 162)', fontWeight: 'bold' }}>
                  Add
                  {' '}
                  {titleName}
                </span>
              </button>
            </div>
          )
        }
        {
          (items.length > 0 && canAdd) && (
            <div className="custom-array-template-add-btn" data-testid="custom-array-template-add-btn" style={{ marginTop: -10 }}>
              <button
                className="btn"
                type="button"
                onClick={(e) => { onAddClick(e); editor.setArrayField(items.length) }}
              >
                <span style={{ color: 'rgb(31, 107, 162)', marginLeft: '-12px' }}>
                  <i className="fa-solid fa-circle-plus fa-2xl" />
                </span>
                {' '}
                <span style={{ color: 'rgb(31, 107, 162)', fontWeight: 'bold' }}>
                  Add another
                  {' '}
                  {titleName}
                </span>
              </button>
            </div>
          )
        }
      </div>
    )
  }
}
export default observer(CustomArrayTemplate)
