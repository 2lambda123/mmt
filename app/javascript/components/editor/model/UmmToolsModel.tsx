/* eslint-disable @typescript-eslint/no-explicit-any */
import toolInformationUiSchema from '../data/configuration/uiSchemas/tools/tool_information'
import relatedUrlsUiSchema from '../data/configuration/uiSchemas/tools/related_urls'
import toolsConfiguration from '../data/configuration/uiForms/toolsConfiguration'
import organizationUiSchema from '../data/configuration/uiSchemas/tools/organization'
import UmmModel from './UmmModel'
import descriptiveKeywordsUiSchema from '../data/configuration/uiSchemas/tools/descriptive_keywords'
import potentialActionUiSchema from '../data/configuration/uiSchemas/tools/potential_action'

export default class UmmToolsModel extends UmmModel {
  get documentTypeForDisplay() {
    return 'Tool Record'
  }
  get documentType() {
    return 'tool_drafts'
  }
  get formSections() {
    return toolsConfiguration
  }

  get uiSchema() {
    const base = { 'ui:submitButtonOptions': { norender: true, submitText: 'Save' } }

    if (this.currentSection.displayName === 'Tool Information') {
      const uiSchema: any = toolInformationUiSchema
      const urlUiSchema: any = uiSchema.URL
      urlUiSchema['ui:service'] = this.service
      urlUiSchema['ui:keyword_scheme'] = 'rucontenttype'
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Descriptive Keywords') {
      const uiSchema = descriptiveKeywordsUiSchema
      const toolKeywords: any = uiSchema.ToolKeywords
      toolKeywords['ui:service'] = this.service
      toolKeywords['ui:keyword_scheme'] = 'science_keywords'
      return uiSchema
    }
    if (this.currentSection.displayName === 'Tool Organizations') {
      const uiSchema: any = organizationUiSchema
      const organizations: any = uiSchema.Organizations.items
      organizations['ui:service'] = this.service
      organizations['ui:keyword_scheme'] = 'providers'
      organizations['ui:keyword_scheme_column_names'] = ['short_name', 'long_name', 'url']
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Potential Action') {
      const uiSchema: any = potentialActionUiSchema
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Related URLs') {
      const uiSchema: any = relatedUrlsUiSchema
      const urlUiSchema: any = uiSchema.RelatedURLs.items
      urlUiSchema['ui:service'] = this.service
      return { ...uiSchema, ...base }
    }
    return { ...base }
  }

  // In this case the old preview sends the same section name that the react component uses (just lowercase), so
  // nothing needs to be migrated. once the old MMT preview is replaced, this 'migrate" functionality can be removed.
  migratedSectionName(sectionName: string): string {
    return sectionName
  }
}
