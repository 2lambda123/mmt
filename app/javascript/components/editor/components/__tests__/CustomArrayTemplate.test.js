import Form from '@rjsf/bootstrap-4'
import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import CustomArrayTemplate from '../CustomArrayTemplate'
import UmmToolsModel from '../../model/UmmToolsModel'
import MetadataEditor from '../../MetadataEditor'

const model = new UmmToolsModel()
const editor = new MetadataEditor(model)
const schema = {
  title: 'CustomArrayFieldTest',
  type: 'array',
  items: {
    properties: {
      firstName: {
        title: 'First Name',
        type: 'string'
      },
      lastName: {
        title: 'Last Name',
        type: 'string'
      },
      age: {
        title: 'Age',
        type: 'integer'
      }
    }
  }
}

CustomArrayTemplate.defaultProps = { options: { editor } }

describe('Custom Array Template', () => {
  it('renders the custom array template and adds an array', async () => {
    const { getByTestId, container } = render(
      <BrowserRouter>
        <Form schema={schema} ArrayFieldTemplate={CustomArrayTemplate} />
      </BrowserRouter>
    )

    const addNewField = getByTestId('custom-array-template-add-btn').querySelector('button[type="button"]')

    fireEvent.click(await addNewField)

    expect(screen.getByTestId('custom-array-element')).toHaveTextContent('First Name')
    expect(screen.getByTestId('custom-array-element')).toHaveTextContent('Last Name')
    expect(screen.getByTestId('custom-array-element')).toHaveTextContent('Age')
    expect(container).toMatchSnapshot()
  })
  it('renders the custom array template. 1.) Adds a field 2.) removes the added field', async () => {
    const { getByTestId, container } = render(
      <BrowserRouter>
        <Form schema={schema} ArrayFieldTemplate={CustomArrayTemplate} />
      </BrowserRouter>
    )
    const addNewField = getByTestId('custom-array-template-add-btn').querySelector('button[type="button"]')

    fireEvent.click(await addNewField)
    expect(screen.getByTestId('custom-array-element')).toHaveTextContent('First Name')
    expect(screen.getByTestId('custom-array-element')).toHaveTextContent('Last Name')
    expect(screen.getByTestId('custom-array-element')).toHaveTextContent('Age')

    const removeField = getByTestId('custom-array-template').querySelector('button[type="button"]')

    fireEvent.click(await removeField)
    expect(container).toMatchSnapshot()
  })
})
