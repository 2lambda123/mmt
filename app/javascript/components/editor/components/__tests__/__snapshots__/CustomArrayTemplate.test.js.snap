// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Array Template renders the custom array template and adds an array 1`] = `
<div>
  <form
    class="rjsf"
  >
    <div
      class="form-group field field-array"
    >
      <div
        class="form-group"
      >
        <div
          data-testid="custom-array-template"
        >
          <div
            class="title"
            style="padding-bottom: 14px; padding-top: 14px;"
          >
            CustomArrayFieldTest
          </div>
          <p />
          <br />
          <div
            class="array-item"
            style="border-bottom: 2px solid rgb(240,240,240); margin-bottom: 25px; padding-left: 10px;"
          >
            <div
              class="custom-array-template-remove-btn"
            >
              <div
                style="display: flex; align-items: center; justify-content: space-between; margin-right: -12px; border-bottom: 2px solid rgb(240,240,240); margin-bottom: 10px; padding-bottom: 10px;"
              >
                <h6
                  style="height: 14px;"
                >
                  <b>
                    CustomArrayFieldTest
                  </b>
                  <span
                    style="color: gray; margin-left: 5px; padding-top: 20px;"
                  >
                    (1 of 1)
                  </span>
                </h6>
                <button
                  class="btn"
                  type="button"
                >
                  <span
                    style="color: red;"
                  >
                    <i
                      class="fa-solid fa-circle-minus fa-2xl"
                    />
                  </span>
                   
                  <span
                    style="color: red; font-weight: bold;"
                  >
                     
                    Remove
                  </span>
                </button>
              </div>
              <div
                class="mb-2  d-flex align-items-center row"
              >
                <div
                  class="array-item col-lg-12 col-12"
                  data-testid="custom-array-element"
                >
                  <div
                    class="form-group field field-undefined"
                  >
                    <div
                      class="form-group"
                    >
                      <div
                        class="p-0 container-fluid"
                      >
                        <div
                          class="row"
                          style="margin-bottom: 10px;"
                        >
                          <div
                            class="col-12"
                          >
                             
                            <div
                              class="form-group field field-string"
                            >
                              <div
                                class="form-group"
                              >
                                <label
                                  class="form-label"
                                  for="root"
                                >
                                  First Name
                                </label>
                                <input
                                  aria-describedby="root__error root__description root__help"
                                  class="form-control"
                                  id="root"
                                  name="root"
                                  placeholder=""
                                  type="text"
                                  value=""
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="row"
                          style="margin-bottom: 10px;"
                        >
                          <div
                            class="col-12"
                          >
                             
                            <div
                              class="form-group field field-string"
                            >
                              <div
                                class="form-group"
                              >
                                <label
                                  class="form-label"
                                  for="root"
                                >
                                  Last Name
                                </label>
                                <input
                                  aria-describedby="root__error root__description root__help"
                                  class="form-control"
                                  id="root"
                                  name="root"
                                  placeholder=""
                                  type="text"
                                  value=""
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="row"
                          style="margin-bottom: 10px;"
                        >
                          <div
                            class="col-12"
                          >
                             
                            <div
                              class="form-group field field-integer"
                            >
                              <div
                                class="form-group"
                              >
                                <label
                                  class="form-label"
                                  for="root"
                                >
                                  Age
                                </label>
                                <input
                                  aria-describedby="root__error root__description root__help"
                                  class="form-control"
                                  id="root"
                                  name="root"
                                  placeholder=""
                                  step="1"
                                  type="number"
                                  value=""
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="custom-array-template-add-btn"
            data-testid="custom-array-template-add-btn"
            style="margin-top: -10px;"
          >
            <button
              class="btn"
              type="button"
            >
              <span
                style="color: rgb(31, 107, 162); margin-left: -12px;"
              >
                <i
                  class="fa-solid fa-circle-plus fa-2xl"
                />
              </span>
               
              <span
                style="color: rgb(31, 107, 162); font-weight: bold;"
              >
                Add another
                 
                CustomArrayFieldTest
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`Custom Array Template renders the custom array template. 1.) Adds a field 2.) removes the added field 1`] = `
<div>
  <form
    class="rjsf"
  >
    <div
      class="form-group field field-array"
    >
      <div
        class="form-group"
      >
        <div
          data-testid="custom-array-template"
        >
          <div
            class="title"
            style="padding-bottom: 14px; padding-top: 14px;"
          >
            CustomArrayFieldTest
          </div>
          <p />
          <br />
          <div
            class="custom-array-template-add-btn"
            data-testid="custom-array-template-add-btn"
            style="margin-top: -30px;"
          >
            <button
              class="btn"
              type="button"
            >
              <span
                style="color: rgb(31, 107, 162); margin-left: -12px;"
              >
                <i
                  class="fa-solid fa-circle-plus fa-2xl"
                />
              </span>
               
              <span
                style="color: rgb(31, 107, 162); font-weight: bold;"
              >
                Add
                 
                CustomArrayFieldTest
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
